File: pwa/src/gitProvider/commitMessages.ts
const normalizeValue = (value: string): string => value.trim().replace(/\s+/g, ' ');

const truncate = (value: string, maxLength: number): string =>
  value.length > maxLength ? value.slice(0, maxLength).trimEnd() : value;

const normalizeMapName = (mapName: string): string =>
  truncate(normalizeValue(mapName), 48) || 'map';

export const buildTodoAddCommitMessage = (shortText: string): string => {
  const normalized = truncate(normalizeValue(shortText), 72);
  return `todo:add ${normalized}`;
};

export const buildTodoEditCommitMessage = (id: string): string =>
  `todo:edit ${normalizeValue(id)}`;

export const buildTodoToggleCommitMessage = (
  id: string,
  completed: boolean,
): string => `todo:toggle ${normalizeValue(id)} -> ${completed ? 'done' : 'open'}`;

export const buildTodoDeleteCommitMessage = (id: string): string =>
  `todo:delete ${normalizeValue(id)}`;

export const buildMapRenameCommitMessage = (oldMapName: string, newMapName: string): string =>
  `map:rename ${normalizeMapName(oldMapName)} -> ${normalizeMapName(newMapName)}`;

export const buildNodeTaskStateCommitMessage = (mapName: string, nodeId: string, taskState: number): string => {
  const stateLabel =
    taskState === 1 ? 'todo' :
      taskState === 2 ? 'doing' :
        taskState === 3 ? 'done' :
          'clear';
  return `map:task ${normalizeMapName(mapName)} ${normalizeValue(nodeId)} -> ${stateLabel}`;
};

export const buildNodeHideDoneTasksCommitMessage = (mapName: string, nodeId: string, hideDoneTasks: boolean): string =>
  `map:hide-done ${normalizeMapName(mapName)} ${normalizeValue(nodeId)} -> ${hideDoneTasks ? 'hide' : 'show'}`;

export const buildNodeStarCommitMessage = (mapName: string, nodeId: string, starred: boolean): string =>
  `map:star ${normalizeMapName(mapName)} ${normalizeValue(nodeId)} -> ${starred ? 'starred' : 'unstarred'}`;

export const buildAttachmentAddCommitMessage = (mapName: string, fileName: string): string =>
  `map:attach ${normalizeMapName(mapName)} ${truncate(normalizeValue(fileName), 48)}`.trimEnd();

export const buildAttachmentRemoveCommitMessage = (mapName: string, fileName: string): string =>
  `map:detach ${normalizeMapName(mapName)} ${truncate(normalizeValue(fileName), 48)}`.trimEnd();

File: android/app/src/main/java/com/systemssanity/focus/domain/maps/CommitMessages.kt
package com.systemssanity.focus.domain.maps

object CommitMessages {
    fun nodeAdd(mapName: String, shortText: String, kind: String): String =
        "map:add ${if (kind == "task") "task" else "note"} ${mapName.normalizedMapName()} ${shortText.normalized(48)}".trimEnd()

    fun nodeEdit(mapName: String, nodeId: String): String =
        "map:edit ${mapName.normalizedMapName()} ${nodeId.normalized()}"

    fun nodeTaskState(mapName: String, nodeId: String, taskStateWireValue: Int): String {
        val label = when (taskStateWireValue) {
            1 -> "todo"
            2 -> "doing"
            3 -> "done"
            else -> "clear"
        }
        return "map:task ${mapName.normalizedMapName()} ${nodeId.normalized()} -> $label"
    }

    fun nodeHideDone(mapName: String, nodeId: String, hideDoneTasks: Boolean): String =
        "map:hide-done ${mapName.normalizedMapName()} ${nodeId.normalized()} -> ${if (hideDoneTasks) "hide" else "show"}"

    fun nodeDelete(mapName: String, nodeId: String): String =
        "map:delete ${mapName.normalizedMapName()} ${nodeId.normalized()}"

    fun mapCreate(mapName: String): String =
        "map:create ${mapName.normalizedMapName()}"

    fun mapDelete(mapName: String): String =
        "map:drop ${mapName.normalizedMapName()}"

    fun mapRename(oldMapName: String, newMapName: String): String =
        "map:rename ${oldMapName.normalizedMapName()} -> ${newMapName.normalizedMapName()}"

    fun conflictResolve(mapName: String): String =
        "map:resolve ${mapName.normalizedMapName()}"

    fun attachmentAdd(mapName: String, fileName: String): String =
        "map:attach ${mapName.normalizedMapName()} ${fileName.normalized(48)}".trimEnd()

    fun attachmentRemove(mapName: String, fileName: String): String =
        "map:detach ${mapName.normalizedMapName()} ${fileName.normalized(48)}".trimEnd()

    private fun String.normalized(maxLength: Int = Int.MAX_VALUE): String =
        trim().replace(Regex("\\s+"), " ").let { if (it.length > maxLength) it.take(maxLength).trimEnd() else it }

    private fun String.normalizedMapName(): String =
        normalized(48).ifBlank { "map" }
}
